lab2/lab2.md

```diff
@@ -1,30 +1,624 @@
+| opcode  |opcode format| branch | memread | toreg |   add  | memwrite | immediate | regwrite | alusrc1 | jump |
+|---------|-------------|--------|---------|-------|--------|----------|-----------|----------|---------|------|
+|    -    |   default   | false  |  false  |   3   |  false |  false   |   false   |   false  |    0    |   0  |
+| 0000000 |   invalid   | false  |  false  |   0   |  false |  false   |   false   |   false  |    0    |   0  |
+| 0110011 |      R      | false  |  false  |   0   |  false |  false   |   false   |   true   |    0    |   0  |
+| 0010011 |      I      | false  |  false  |   0   |  false |  false   |   true    |   true   |    0    |   0  |
+| 0000011 |    load     | false  |  true   |   1   |  true  |  false   |   true    |   true   |    0    |   0  |
+| 0100011 |    store    | false  |  false  |   0   |  true  |  true    |   true    |   false  |    0    |   0  |
+| 1100011 |    beq      | true   |  false  |   0   |  false |  false   |   false   |   false  |    0    |   0  |
+| 0110111 |    lui      | false  |  false  |   0   |  true  |  false   |   true    |   true   |    1    |   0  |
+| 0010111 |    auipc    | false  |  false  |   0   |  true  |  false   |   true    |   true   |    2    |   0  |
+| 1101111 |    jal      | true   |  false  |   2   |  false |  false   |   false   |   true   |    1    |   2  |
+| 1100111 |    jalr     | false  |  false  |   2   |  false |  false   |   true    |   true   |    0    |   3  |
+
+
+
+Given the input opcode, you must generate the correct control signals.
+The template code from `src/main/scala/components/control.scala` is shown below.
+You will fill in where it says *Your code goes here*.
+```
+// Control logic for the processor
+
+package dinocpu
+
+import chisel3._
+import chisel3.util.{BitPat, ListLookup}
+
+/**
+ * Main control logic for our simple processor
+ *
+ * Output: branch,  true if branch or jal and update PC with immediate
+ * Output: memread, true if we should read memory
+ * Output: toreg, 0 if writing ALU result, 1 if writing memory data, 2 if writing pc+4
+ * Output: add, true if the ALU should add the results
+ * Output: memwrite, write the memory
+ * Output: regwrite, write the register file
+ * Output: immediate, true if use the immediate value
+ * Output: alusrc1, 0 is Read data 1, 1 is zero, 2 is PC
+ * Output: jump, 0 no jump, 2 jump, 3 jump and link register
+ *
+ * For more information, see section 4.4 of Patterson and Hennessy
+ * This follows figure 4.22
+ */
+class Control extends Module {
+  val io = IO(new Bundle {
+    val opcode = Input(UInt(7.W))
+
+    val branch = Output(Bool())
+    val memread = Output(Bool())
+    val toreg = Output(UInt(2.W))
+    val add = Output(Bool())
+    val memwrite = Output(Bool())
+    val regwrite = Output(Bool())
+    val immediate = Output(Bool())
+    val alusrc1 = Output(UInt(2.W))
+    val jump    = Output(UInt(2.W))
+  })
+
+  val signals =
+    ListLookup(io.opcode,
```

Comment:
We need to explain this syntax in detail so they can use it. Explain how it is just a copy of the table above. Also explain how the opcode is checked against the bit pattern in the first entry and then the signals are in order.